tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+}
+
+// compactionFileName will return the file name that should be used for a conmpaction onthis partition
+func (p *partition) compactionFileName() string {
+	return filepath.Join(p.path, fmt.Sprintf("%02d.%06d.CPT", p.id, 1))
+}
+
+// fileIDFromName will return the segment ID from the file name
+func (p *partition) fileIDFromName(name string) (uint32, error) {
+	parts := strings.Split(filepath.Base(name), ".")
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("file name doesn't follow wal format: %s", name)
+	}
+	id, err := strconv.ParseUint(parts[1], 10, 32)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(id), nil
+}
+
+// shouldFlush returns true if a partition should be flushed. The criteria are
+// maxSeriesSize - flush if any series in the partition has exceeded this size threshold
+// readySeriesSize - a series is ready to flush once it has this much data in it
+// compactionThreshold - a partition is ready to flush if this percentage of series has hit the readySeriesSize or greater
+// partitionSizeThreshold - a partition should be flushed if it has reached this size in memory
+func (p *partition) shouldFlush(maxSeriesSize, readySeriesSize int, compactionThreshold float64, partitionSizeThreshold uint64) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.totalSize > partitionSizeThreshold {
+		return true
+	}
+
+	countReady := float64(0)
+	for _, s := range p.cacheSizes {
+		if s > maxSeriesSize {
+			return true
+		} else if s > readySeriesSize {
+			countReady += 1
+		}
+	}
+
+	if countReady/float64(len(p.cacheSizes)) > compactionThreshold {
+		return true
+	}
+
+	return false
+}
+
+// flushAndCompact will flush any series that are over their threshold and then read in all old segment files and
+// write the data that was not flushed to a new file
+func (p *partition) flushAndCompact(idx indexWriter, maxSeriesSize, readySeriesSize int) error {
+	seriesToFlush := make(map[string][][]byte)
+	sizeOfFlush := 0
+	var compactFilesLessThan uint32
+
+	if err := func() error {
+		p.mu.Lock()
+		defer p.mu.Unlock()
+
+		if p.compactionRunning {
+			return ErrCompactionRunning
+		}
+		p.compactionRunning = true
+
+		for k, s := range p.cacheSizes {
+			// if the series is over the threshold, save it in the map to flush later
+			if s >= maxSeriesSize || s >= readySeriesSize {
+				sizeOfFlush += s
+				seriesToFlush[k] = p.cache[k]
+				delete(p.cacheSizes, k)
+				delete(p.cache, k)
+
+				// always hand the index data that is sorted
+				if p.cacheDirtySort[k] {
+					sort.Sort(byteSlices(seriesToFlush[k]))
+					delete(p.cacheDirtySort, k)
+				}
+			}
+		}
+		p.flushCache = seriesToFlush
+
+		// roll over a new segment file so we can compact all the old ones
+		if err := p.newSegmentFile(); err != nil {
+			return err
+		}
+		compactFilesLessThan = p.currentSegmentID
+
+		return nil
+	}(); err == ErrCompactionRunning {
+		return nil
+	} else if err != nil {
+		return err
+	}
```

Comment:
Can you move this inline function to a for realz function? `flushAndCompact()` is getting huge.